### PR TITLE
SNOW-1447365: Add support for to_csv

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1447365

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

Add support for DataFrame.to_csv and Series.to_csv.
Also supports writing to snowflake stage location.

### CODE REVIEW

File: tests/integ/modin/io/test_to_csv.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    # Write csv with native pandas.
+    native_series.to_csv(native_path, **kwargs)
+    # Write csv to snowflake stage.
+
+    stage_location = f"@{sf_stage}/{os.path.basename(snow_path)}"
+    if kwargs.get("index", True) is False:
+        # Writing single column with null values has different behavior in snowpark
+        # pandas as compared to native pandas. Null values are written as empty string
+        # in snowpark pandas but in native pandas null values are written as a pair of
+        # duoblequotes "". Add na_rep param to match behavior in test.
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/io/test_to_csv.py:140]:
```suggestion
        # double quotes "". Add na_rep param to match behavior in test.
```

File: tests/integ/modin/io/test_to_csv.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    kwargs = kwargs.copy()
+    kwargs.pop("ext", None)
+    # Write csv with native pandas.
+    native_series.to_csv(native_path, **kwargs)
+    # Write csv to snowflake stage.
+
+    stage_location = f"@{sf_stage}/{os.path.basename(snow_path)}"
+    if kwargs.get("index", True) is False:
+        # Writing single column with null values has different behavior in snowpark
```
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/io/test_to_csv.py:137]: why not match pandas behavior?

**Review by sfc-gh-mvashishtha — COMMENTED** [review]
I looked at src/snowflake/snowpark/modin/plugin/utils/error_message.py and tests/integ/modin/io/test_to_csv.py, but I still have to review the rest of the files.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 3]
```diff
@@ -4350,3 +4350,159 @@ def merge():
...
+    def to_csv():
+        """
+        Write object to a comma-separated values (csv) file. This can write csv file
+        either to local filesystem or to snowflake stage. Filepath staring with `@` is
+        treated as snowflake stage location.
+
+        Note: Writing to local filesystem supports all parameters but writing to
+        snowflake stage does not support float_format, mode, encoding, quoting,
+        quotechar, lineterminator, doublequote and decimal parameters. Also when
+        writing to snowflake stage chucksize, errors and storage_options parameters
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4363]:
```suggestion
        writing to snowflake stage, chunksize, errors, and storage_options parameters
```

File: tests/integ/modin/io/test_to_csv.py (hunk 2 above)
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/io/test_to_csv.py:137]: Maybe the pandas behavior is a bug. I filed https://github.com/pandas-dev/pandas/issues/59116

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 4]
```diff
@@ -4350,3 +4350,159 @@ def merge():
...
+            As an example, the following could be passed for faster compression and to create
+            a reproducible gzip archive:
+            ``compression={'method': 'gzip', 'compresslevel': 1, 'mtime': 1}``.
+
+               May be a dict with key 'method' as compression mode
+               and other entries as additional compression options if
+               compression mode is 'zip'.
+
+               Passing compression options as keys in dict is
+               supported for compression modes 'gzip', 'bz2', 'zstd', and 'zip'.
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4426]: is this portion repeating information from the first part of the `compression` description?

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py [hunk 5]
```diff
@@ -3268,3 +3268,155 @@ def str():
...
+        --------
+        read_csv : Load a CSV file into a DataFrame.
+        to_excel : Write DataFrame to an Excel file.
+
+        Examples
+        --------
+        Create 'out.csv' containing 'series' without indices
+
+        >>> series = pd.Series(['red', 'green', 'blue'], name='color')
+        >>> series.to_csv('out.csv', index=False)  # doctest: +SKIP
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/series.py:3400]: why skip doctests here?

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py [hunk 6]
```diff
@@ -3268,3 +3268,155 @@ def str():
...
+            to one of {``'zip'``, ``'gzip'``, ``'bz2'``, ``'zstd'``, ``'xz'``, ``'tar'``} and
+            other key-value pairs are forwarded to
+            ``zipfile.ZipFile``, ``gzip.GzipFile``,
+            ``bz2.BZ2File``, ``zstandard.ZstdCompressor``, ``lzma.LZMAFile`` or
+            ``tarfile.TarFile``, respectively.
+            As an example, the following could be passed for faster compression and to create
+            a reproducible gzip archive:
+            ``compression={'method': 'gzip', 'compresslevel': 1, 'mtime': 1}``.
+
+               May be a dict with key 'method' as compression mode
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/series.py:3337]: ditto about repeating the compression dict description.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 7]
```diff
@@ -4350,3 +4350,159 @@ def merge():
...
+        to_excel : Write DataFrame to an Excel file.
+
+        Examples
+        --------
+        Create 'out.csv' containing 'df' without indices
+
+        >>> df = pd.DataFrame({{'name': ['Raphael', 'Donatello'],
+        ...                    'mask': ['red', 'purple'],
+        ...                    'weapon': ['sai', 'bo staff']}})
+        >>> df.to_csv('out.csv', index=False)  # doctest: +SKIP
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4486]: why skip doctests here?

**Review by sfc-gh-mvashishtha — COMMENTED** [review]
I've gone through all the files once!

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 4 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4426]: looks repetitive, let me remove this.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 3 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4363]: Ack

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 7 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4486]: These tests are dependent on setting up local filesystem. We can make them work by setting up a temp directory but examples will not look as nicer. So I kept these as it is while copying from pandas doc. Let me know what do you think?

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py (hunk 6 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/series.py:3337]: removed

File: tests/integ/modin/io/test_to_csv.py (hunk 2 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/io/test_to_csv.py:137]: Yeah, this behavior seems very off to me. Thanks for filing the bug.

File: tests/integ/modin/io/test_to_csv.py (hunk 1 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/io/test_to_csv.py:140]: Ack

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 7 above)
Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4602]: I didn't think about the fact that these doctests create new files. I'm okay with skipping the doctests.

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py (hunk 5 above)
Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/docstrings/series.py:3434]: resolved here: https://github.com/snowflakedb/snowpark-python/pull/1832#discussion_r1657696174

File: tests/integ/modin/io/test_to_csv.py (hunk 2 above)
Comment (sfc-gh-mvashishtha) [reply tests/integ/modin/io/test_to_csv.py:137]: Could you add a link to the pandas bug here? I'm still not sure whether it's a bug, but we can revisit the behavior later.

Comment (sfc-gh-nkumar) [reply tests/integ/modin/io/test_to_csv.py:137]: Added link to bug.

**Review by sfc-gh-mvashishtha — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 8]
```diff
@@ -4473,3 +4473,152 @@ def merge():
         2  bar      7
         3  bar      8
         """
+
+    def to_csv():
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4477]: Maybe dumb question; but why does this not have parameters in the signature?

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:4477]: This is a dummy dataframe just to override docstrings.

**Review by sfc-gh-mvashishtha — APPROVED** [review]

**Review by sfc-gh-jkew — APPROVED** [review]

**Pull request merged**